libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/internal/kapt/KaptWithoutKotlincTask.kt

```diff
@@ -201,29 +216,39 @@ abstract class KaptWithoutKotlincTask @Inject constructor(private val workerExec
 private class KaptExecution @Inject constructor(
     val optionsForWorker: KaptOptionsForWorker,
     val toolsJarURLSpec: String,
-    val kaptClasspath: List<File>
+    val kaptClasspath: List<File>,
+    val classloadersCacheSize: Int
 ) : Runnable {
     private companion object {
         private const val JAVAC_CONTEXT_CLASS = "com.sun.tools.javac.util.Context"
 
         private fun kaptClass(classLoader: ClassLoader) = Class.forName("org.jetbrains.kotlin.kapt3.base.Kapt", true, classLoader)
-        private var cachedClassLoaderWithToolsJar: ClassLoader? = null
+
+        private var classLoadersCache: ClassLoadersCache? = null
+
         private var cachedKaptClassLoader: ClassLoader? = null
     }
 
+    private val logger = LoggerFactory.getLogger(KaptExecution::class.java)
+
     override fun run(): Unit = with(optionsForWorker) {
         val kaptClasspathUrls = kaptClasspath.map { it.toURI().toURL() }.toTypedArray()
         val rootClassLoader = findRootClassLoader()
 
-        val classLoaderWithToolsJar = cachedClassLoaderWithToolsJar ?: if (!toolsJarURLSpec.isEmpty() && !javacIsAlreadyHere()) {
-            URLClassLoader(arrayOf(URL(toolsJarURLSpec)), rootClassLoader)
-        } else {
-            rootClassLoader
+        if (cachedKaptClassLoader == null) {
+            val classLoaderWithToolsJar = if (toolsJarURLSpec.isNotEmpty() && !javacIsAlreadyHere()) {
+                URLClassLoader(arrayOf(URL(toolsJarURLSpec)), rootClassLoader)
+            } else {
+                rootClassLoader
+            }
+            cachedKaptClassLoader = URLClassLoader(kaptClasspathUrls, classLoaderWithToolsJar)
         }
-        cachedClassLoaderWithToolsJar = classLoaderWithToolsJar
+        val kaptClassLoader = cachedKaptClassLoader!!
```

Comment:
Note: you can get rid of `!!` by writing something like

```kotlin
cachedKaptClassLoader ? run {
    val classLoaderWithToolsJar = if (toolsJarURLSpec.isNotEmpty() && !javacIsAlreadyHere()) {
        URLClassLoader(arrayOf(URL(toolsJarURLSpec)), rootClassLoader)
    } else {
        rootClassLoader
    }
    URLClassLoader(kaptClasspathUrls, classLoaderWithToolsJar)
}
```

Reply:
ok, I'll refactor it so it won't need !!